### DISCUSSION TITLE
How can I get handle from a control not window?

### DISCUSSION THREAD

**Discussion by fallingrust** [microsoft/WindowsAppSDK#1760] [Q&A]
It is possible that I get handle from a control not window in WinUI3? I have a WinForm  application which use SDL 2、ffmpeg (a native lib)to render some network stream with control‘s hwnd.It is possible that I can refactoring  this project,but do not change the native lib?

**Comment by andrewleader** [comment]
*EDIT: The original question was edited, originally it was asking how to get a handle from a WinForm app, but now it was updated to ask how to get a handle from a WinUI 3 control*

Hi @fallingrust, it looks like WinForms controls have a `Handle` property that contains the HWND: https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.control.handle?view=windowsdesktop-5.0

For example...

```csharp
var hwnd = Button2.Handle;
```

**Reply by fallingrust** [reply]
Yes，I know,but It is possible that get handle with WinUI3? Or some other way，like wpf host winform control?

**Reply by jonwis** [reply]
@MikeHillberg or @jevansaks might help redirect as well.

@fallingrust - so you're looking to have an animated button playing a streamed network video?

**Reply by fallingrust** [reply]
yes, in winform I send the handle to the native library for video rendering,and I think it is a good choice to change WinForm to WinUI3 without modifying the native library, beautiful UI and high performance.But the most important thing is not to modify the native library, after all, I'm not an expert in C++/C.

**Reply by jevansaks** [reply]
In a WinUI3 window, the content is all HWND-less and the whole scene just has one HWND at the top. Generally libraries like that will want the HWND so they can create a DX swapchain and start rendering to it, but that's not the best way to do it with WinUI. Does the library have a way to render "windowless"? If so you can then use [SwapChainPanel](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.swapchainpanel?view=winui-3.0) to get that content into the scene.

**Reply by fallingrust** [reply]
Thanks for your answer, I need to consult some documents about this.  :)

**Reply by jevansaks** [reply]
Can you share a link to the API documentation for the library you're using? I might be able to help point you in the right direction.

**Reply by fallingrust** [reply]
Sorry,it is just an unofficial lib,coded by my friend.And I need talk with him about  swapchain and SwapChainPanel. Thank you again for your reply. :-)

**Comment by castorix** [comment]
You can create a Win32 control as child and then get its handle
I tested with an ATL control to host a WMP control and it works (I used a region (SetWindowRgn) so that it appears through the "Microsoft.UI.Content.ContentWindowSiteBridge" window)
